jep/0000/README.adoc

```diff
@@ -0,0 +1,298 @@
+Use of these annotations takes precedence over the presence of the layout tags listed in the above subsection.
+
+This provides methods to specify and override default views and fragments.
+For example, suppose Jenkins defines an abstract class `BaseClass` which has views `index`, `configure`, `create`, and `delete`.
+Implementation classes are expected to provide their own fragment view `config` which gets included in `configure` and `create`.
+This can be encoded explicitly in the model class:
+
+[source,java]
+----
+@StaplerViews({"configure", "create", "delete"}) // index is implicit
+@StaplerFragments("config")
+public abstract class BaseClass {
+    ...
+}
+----
+
+By adding the annotations to Jenkins, implementation classes (typically from plugins) can be filtered explicitly via the parent annotations without updating the downstream code.
+Implementation classes may further define additional views or fragments.
+For example, suppose `ExtendedClass` extends `BaseClass` and adds its own optional fragment view `advanced` which its implementation classes can include.
+
+[source,java]
+----
+@StaplerFragments("advanced")
+public class ExtendedClass extends BaseClass {
+    ...
+}
+----
+
+=== Whitelist File
+
+Jenkins ships with a default whitelist of common, known views in plugins that would otherwise be broken by the above rules.
+This file consists of class names and their respective views and fragments.
+These entries have the same effect as annotating the listed classes with the above annotations.
+The syntax for this file is defined below.
+
+[source,bnf]
+----
+<line> ::= <fully-qualified-model-class-name> <ws> <view-list>
+<view-list> ::= <view-entry> | <view-entry> <ws> <view-list>
+<view-entry> ::= <view-name> | "!" <view-name>
+<ws> ::= " " | " " <ws>
+----
+
+For each view entry, if the view name is prefixed by a `!`, then it is added as a fragment view.
+Otherwise, the view name is added as a dispatchable view.
+
+Additional whitelist entries can be specified by the file `$JENKINS_HOME/stapler-views-whitelist.txt` or whatever path is specified by the system property `jenkins.security.stapler.StaplerDispatchValidator.whitelist`.
+
+== Motivation
+
+During exploratory testing, the Jenkins security team discovered a class of vulnerabilities in Jenkins.
+Common to all these vulnerabilities is the fact that Stapler does not make a distinction between views that are meant to be dispatched directly versus those meant to only be included in other views.
+This lack of distinction causes numerous problems in Jenkins where URLs can be constructed to directly access fragment views of models which oftentimes do not perform their own permission checks or other initialization logic.
+It was discovered through analysis of existing well-behaved views that they all had implicit or explicit use of page initialization logic including setting an response `Content-Type`.
```

Comment:
This could be clarified a little bit. It isn't clear what the part "setting a response `Content-Type`" is connected to. The next sentence talks about implicit behavior. The end of this sentence seems to be talking about explicit behavior, but isn't clear.

Perhaps break it up to something like:
It was discovered through analysis of existing well-behaved views that they all had implicit or explicit use of page initialization logic. Explicit initialization logic includes setting a response `Content-Type`. Most views use implicit initialization through `l:layout` or `l:ajax` Jelly tags.


Maybe. I'm still not sure the best way to get the idea across.